File: ctci/tmp/chapter_10_sorting_searching/js/10.11-peaksAndValleys.js
/**
 * Sorts an array into an alternating sequence of peaks and valleys.
 * @param   {number[]}   arr
 */
const sortValleyPeak = arr => {
  for (let i = 1; i < arr.length; i += 2) {
    const max = findMax(arr, i - 1, i, i + 1)
    if (max !== i) [arr[i], arr[max]] = [arr[max], arr[i]]
  }
}

/**
 * Finds the max value of three given indexes in an array and returns the
 * associated index.
 * @param   {number[]}  arr
 * @param   {number}    prev
 * @param   {number}    curr
 * @param   {number}    next
 * @returns {number}
 */
const findMax = (arr, prev, curr, next) => {
  const a = arr[prev]
  const b = arr[curr]
  const c = next < arr.length ? arr[next] : Number.MIN_SAFE_INTEGER
  const biggest = Math.max(a, b, c)
  if (biggest === a) return prev
  if (biggest === b) return curr
  return next
}

let arr = [9, 1, 0, 4, 8, 7]
sortValleyPeak(arr)
console.log(arr) // [ 1, 9, 0, 8, 4, 7 ]

arr = [-5, -4, 3]
sortValleyPeak(arr)
console.log(arr) // [ -5, 3, -4 ]

arr = [-2]
sortValleyPeak(arr)
console.log(arr) // [ -2 ]

arr = [1, 1, 1]
sortValleyPeak(arr)
console.log(arr) // [ 1, 1, 1 ]

File: ctci/tmp/chapter_10_sorting_searching/python/10.11-peaks_and_valleys.py
from typing import List


def sort_valley_peak(arr: List):
    """
    Sorts an array into an alternating sequence of peaks and valleys.
    """
    for i in range(1, len(arr), 2):
        m = find_max(arr, i - 1, i, i + 1)
        if i != m:
            arr[i], arr[m] = arr[m], arr[i]


def find_max(arr: List, prev: int, curr: int, next: int) -> int:
    """
    Finds the max value of three given indexes in an array and returns the
    associated index.
    """
    a, b = arr[prev], arr[curr]
    c = arr[next] if next < len(arr) else float("-inf")
    biggest = max(a, b, c)
    if a == biggest:
        return prev
    if b == biggest:
        return curr
    return next


arr = [9, 1, 0, 4, 8, 7]
sort_valley_peak(arr)
print(arr)  # [1, 9, 0, 8, 4, 7

arr = [-5, -4, 3]
sort_valley_peak(arr)
print(arr)  # [-5, 3, -4]

arr = [-2]
sort_valley_peak(arr)
print(arr)  # [-2]

arr = [1, 1, 1]
sort_valley_peak(arr)
print(arr)  # [1, 1, 1]
